src/util.ts

```diff
@@ -181,19 +182,12 @@ export const encodeProposal = (
  * @param encodedProposal - Encoded proposal of the referendum - aka inlined preimage.
  */
 export const getReferendumId = async (
-  apiAtBlock: ApiDecoration<"promise">,
+  api: API,
+  blockHash: string,
   encodedProposal: string,
 ): Promise<undefined | number> => {
-  // https://github.com/paritytech/substrate/blob/63246b699d7e2645c8b12aae46f8f0765c682183/frame/referenda/src/lib.rs#L271-L278
-  // [index, track, proposal]
-  type ReferendumSubmittedData = [number, number, Record<string, unknown>];
-
-  const events = await apiAtBlock.query.system.events();
-  const referendumEvent = events.find(
-    (record) =>
-      record.event.section === "referenda" &&
-      record.event.method === "Submitted" &&
-      (record.event.data.toJSON() as ReferendumSubmittedData)[2].inline === encodedProposal,
-  );
-  return (referendumEvent?.event.data.toJSON() as ReferendumSubmittedData)?.[0];
+  const referendums = await api.event.Referenda.Submitted.pull();
+  const referendum = referendums.filter((r) => r.meta.block.hash === blockHash);
+  // TODO: Find a way to check that the referendum is the same as the encoded proposal
+  return referendum.length > 0 ? referendum[0].payload.track : undefined;
```

Comment:
Should probably return `referendum[0].payload.index` instead?

Reply:
Good catch. Fixing it